src/function/gds/louvain.cpp

```diff
@@ -0,0 +1,483 @@
+        commInfos.assign(numNodes, CommInfo());
+        pastCommunities.assign(numNodes, UNASSIGNED_COMM);
+        currCommunities.assign(numNodes, UNASSIGNED_COMM);
+        targetCommunities.assign(numNodes, UNASSIGNED_COMM);
+        totalWeight = 0.0;
+    }
+
+    void startNewIter() {
+        clusterWeightInternal.assign(graph.numNodes, 0.0);
+        commInfosNext.assign(graph.numNodes, CommInfoAtomic());
+    }
+
+    void insertNewNode(const offset_t nodeId) {
+        graph.insertNewNode();
+        commInfos[nodeId].size = 1;
+        commInfos[nodeId].degree = 0.0;
+        pastCommunities[nodeId] = nodeId;
+        currCommunities[nodeId] = nodeId;
+    }
+
+    void insertEdge(const offset_t from, const offset_t to, const double weight = DEFAULT_WEIGHT) {
+        graph.insertEdge(to, weight);
+        nodeWeightedDegrees[from] += weight;
+        commInfos[from].degree += weight;
+        totalWeight += weight;
+    }
+
+    void finalize() { graph.insertNewNode(); }
+};
+
+struct FinalResults {
+    vector<offset_t> communities;
+
+    explicit FinalResults(const offset_t numNodes) { communities.resize(numNodes); }
+};
+
+class SaveCommAssignmentsVC : public InMemVertexCompute {
+public:
+    explicit SaveCommAssignmentsVC(offset_t phaseId, FinalResults& finalResults, PhaseState& state)
+        : phaseId{phaseId}, finalResults{finalResults}, state{state} {}
+
+    void vertexCompute(offset_t startOffset, offset_t endOffset) override {
+        if (phaseId == 0) {
+            for (auto nodeId = startOffset; nodeId < endOffset; ++nodeId) {
+                finalResults.communities[nodeId] = state.pastCommunities[nodeId];
+            }
+        } else {
+            for (auto nodeId = startOffset; nodeId < endOffset; ++nodeId) {
+                auto prevCommunity = finalResults.communities[nodeId];
+                // Every previous community becomes a node in the current phase.
+                auto newCommunity = state.pastCommunities[prevCommunity];
+                finalResults.communities[nodeId] = newCommunity;
+            }
+        }
+    }
+
+    std::unique_ptr<InMemVertexCompute> copy() override {
+        return std::make_unique<SaveCommAssignmentsVC>(phaseId, finalResults, state);
+    }
+
+private:
+    offset_t phaseId;
+    FinalResults& finalResults;
+    PhaseState& state;
+};
+
+class RunIterationVC : public InMemVertexCompute {
+public:
+    explicit RunIterationVC(PhaseState& state) : state{state} {}
+
+    void vertexCompute(offset_t startOffset, offset_t endOffset) override {
+        double constantForSecondTerm = 1.0 / state.totalWeight;
+        for (auto nodeId = startOffset; nodeId < endOffset; ++nodeId) {
+            auto from = state.graph.nodes[nodeId];
+            auto to = state.graph.nodes[nodeId + 1];
+            double selfLoop = 0.0;
+            unordered_map<offset_t, offset_t> clusterLocalMap;
+            vector<double> counter;
+            if (from != to) {
```

Comment:
I would rewrite the coe to see if you don't have to special case from != to (if it simplifies the code).

Reply:
Code is simpler now